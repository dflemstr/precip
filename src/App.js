import React from 'react'
import Grid from '@material-ui/core/Grid'
import Plant from './Plant'
import LinearProgress from '@material-ui/core/LinearProgress'
import Error from '@material-ui/icons/Error'
import Typography from '@material-ui/core/Typography'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import { withStyles } from '@material-ui/core/styles'

const styles = theme => ({
  root: {
    flexGrow: 1,
    padding: theme.spacing.unit * 3
  },
  error: {
    verticalAlign: 'text-top'
  }
})

class App extends React.Component {
  constructor (props) {
    super(props)
    this.controller = typeof window.AbortController !== 'undefined' ? new window.AbortController() : null
    this.interval = null
    this.state = {error: null, data: null}
    this.mounted = false
  }

  async componentDidMount () {
    this.mounted = true
    this.interval = setInterval(() => this._update(), 60 * 1000)
    await this._update()
  }

  async _update () {
    try {
      const response = await window.fetch('https://s3-eu-west-1.amazonaws.com/precip-stats/data.json', {
        headers: {
          'accept': 'application/json'
        },
        signal: this.controller ? this.controller.signal : null
      })
      const data = await response.json()
      this.onDataReceived(data)
    } catch (error) {
      this.onDataError(error)
    }
  }

  onDataReceived (data) {
    if (this.mounted) {
      this.setState({...this.state, data, error: null})
    }
  }

  onDataError (error) {
    if (this.mounted) {
      this.setState({...this.state, error, data: null})
    }
  }

  componentWillUnmount () {
    if (this.controller) {
      this.controller.abort()
    }

    if (this.interval !== null) {
      clearInterval(this.interval)
      this.interval = null
    }

    this.mounted = false
  }

  render () {
    const {classes} = this.props
    if (this.state.error) {
      return (<main className={classes.root}>
        <Grid container justify='center' spacing={24}>
          <Grid item>
            <Card>
              <CardContent>
                <Typography gutterBottom variant='headline'>
                  <Error className={classes.error} /> Error
                </Typography>
                <Typography>{this.state.error.message}</Typography>
              </CardContent>
            </Card>
          </Grid>
        </Grid>
      </main>)
    } else if (this.state.data) {
      let items = this.state.data.modules.map(module => (<Grid key={module.id} item>
        <Plant title={module.name} module={module} />
      </Grid>))
      return (<main className={classes.root}>
        <Grid container spacing={24}>
          {items}
        </Grid>
      </main>)
    } else {
      return <LinearProgress />
    }
  }
}

export default withStyles(styles)(App)
