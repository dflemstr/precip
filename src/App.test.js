import React from 'react'
import App from './App'
import renderer from 'react-test-renderer'

/* global beforeEach expect fetch it */

beforeEach(() => {
  fetch.resetMocks()
})

it('renders without crashing', (done) => {
  const data = {
    'modules': [{
      'id': '71f3abbf-eda2-4b1f-a9a2-8af7f290e0a6',
      'name': 'Plant 2',
      'running': false,
      'forceRunning': false,
      'historicalMoisture': [{
        'measurementStart': '2018-05-15T17:05:00Z',
        'min': 0.7440120858660881,
        'max': 0.8178992436390712,
        'p25': 0.7769198054604457,
        'p50': 0.7913926075090871,
        'p75': 0.807740966624666
      }]
    }]
  }

  fetch.mockResponseOnce(JSON.stringify(data))

  const component = renderer.create(<App />)

  // TODO: hack
  setTimeout(() => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(fetch.mock.calls[0][0]).toEqual('https://s3-eu-west-1.amazonaws.com/precip-stats/data.json')
    done()
  }, 500)
})

it('renders moisture values', (done) => {
  const data = {
    'modules': [{
      'id': '71f3abbf-eda2-4b1f-a9a2-8af7f290e0a6',
      'name': 'Plant 2',
      'running': false,
      'forceRunning': false,
      'minMoisture': 0.25,
      'maxMoisture': 0.9,
      'lastMoisture': 0.8,
      'historicalMoisture': [{
        'measurementStart': '2018-05-15T17:05:00Z',
        'min': 0.7440120858660881,
        'max': 0.8178992436390712,
        'p25': 0.7769198054604457,
        'p50': 0.7913926075090871,
        'p75': 0.807740966624666
      }]
    }]
  }

  fetch.mockResponseOnce(JSON.stringify(data))

  const component = renderer.create(<App />)

  // TODO: hack
  setTimeout(() => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(fetch.mock.calls[0][0]).toEqual('https://s3-eu-west-1.amazonaws.com/precip-stats/data.json')
    done()
  }, 500)
})
