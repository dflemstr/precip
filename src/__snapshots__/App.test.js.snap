// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error 1`] = `
<main
  className="App-root-1"
>
  <div
    className="MuiGrid-container-19 MuiGrid-spacing-xs-24-42 MuiGrid-justify-xs-center-36"
  >
    <div
      className="MuiGrid-item-20"
    >
      <div
        className="MuiPaper-root-111 MuiPaper-elevation2-115 MuiPaper-rounded-112 MuiCard-root-110"
      >
        <div
          className="MuiCardContent-root-168"
        >
          <h1
            className="MuiTypography-root-144 MuiTypography-headline-149 MuiTypography-gutterBottom-161"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-178 App-error-2"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g>
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                />
              </g>
            </svg>
             Error
          </h1>
          <p
            className="MuiTypography-root-144 MuiTypography-body1-153"
          >
            failure foo bar baz test
          </p>
        </div>
      </div>
    </div>
  </div>
</main>
`;

exports[`renders moisture values 1`] = `
<main
  className="App-root-1"
>
  <div
    className="MuiGrid-container-19 MuiGrid-spacing-xs-24-42"
  >
    <div
      className="MuiGrid-item-20"
    >
      <div
        className="MuiPaper-root-111 MuiPaper-elevation2-115 MuiPaper-rounded-112 MuiCard-root-110"
      >
        <div
          className="MuiCardHeader-root-138"
          subtitle={undefined}
        >
          <div
            className="MuiCardHeader-content-141"
          >
            <span
              className="MuiTypography-root-144 MuiTypography-headline-149 MuiCardHeader-title-142"
            >
              Plant 2
            </span>
          </div>
        </div>
        <div
          className="MuiCardContent-root-168"
        >
          <div
            className="rv-xy-plot "
            style={
              Object {
                "height": "100px",
                "width": "400px",
              }
            }
          >
            <svg
              className="rv-xy-plot__inner"
              height={100}
              onClick={[Function]}
              onDoubleClick={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseMove={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              onWheel={[Function]}
              style={undefined}
              width={400}
            >
              <path
                className="rv-xy-plot__series rv-xy-plot__series--line "
                d="M400,45.129104661107796L400,60Z"
                onClick={[Function]}
                onContextMenu={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "fill": "rgb(222, 232, 255)",
                    "opacity": 1,
                    "stroke": "rgb(222, 232, 255)",
                  }
                }
                transform="translate(0,0)"
              />
              <path
                className="rv-xy-plot__series rv-xy-plot__series--line "
                d="M400,45.31380060682425L400,60Z"
                onClick={[Function]}
                onContextMenu={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "fill": "rgb(197, 208, 242)",
                    "opacity": 1,
                    "stroke": "rgb(197, 208, 242)",
                  }
                }
                transform="translate(0,0)"
              />
              <path
                className="rv-xy-plot__series rv-xy-plot__series--line "
                d="M400,45.61104349983478L400,60Z"
                onClick={[Function]}
                onContextMenu={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "fill": "rgb(172, 183, 229)",
                    "opacity": 1,
                    "stroke": "rgb(172, 183, 229)",
                  }
                }
                transform="translate(0,0)"
              />
              <path
                className="rv-xy-plot__series rv-xy-plot__series--line "
                d="M400,45.87418535526462L400,60Z"
                onClick={[Function]}
                onContextMenu={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "fill": "rgb(146, 159, 216)",
                    "opacity": 1,
                    "stroke": "rgb(146, 159, 216)",
                  }
                }
                transform="translate(0,0)"
              />
              <path
                className="rv-xy-plot__series rv-xy-plot__series--line "
                d="M400,46.472507529707485L400,60Z"
                onClick={[Function]}
                onContextMenu={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "fill": "rgb(121, 134, 203)",
                    "opacity": 1,
                    "stroke": "rgb(121, 134, 203)",
                  }
                }
                transform="translate(0,0)"
              />
              <g
                className="rv-xy-plot__axis rv-xy-plot__axis--horizontal "
                style={
                  Object {
                    "line": Object {
                      "stroke": "none",
                    },
                    "text": Object {
                      "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                      "fontSize": "0.875rem",
                    },
                    "ticks": Object {
                      "fill": "#9e9e9e",
                    },
                  }
                }
                transform="translate(0,60)"
              >
                <line
                  className="rv-xy-plot__axis__line"
                  style={
                    Object {
                      "line": Object {
                        "stroke": "none",
                      },
                      "stroke": "none",
                      "text": Object {
                        "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                        "fontSize": "0.875rem",
                      },
                      "ticks": Object {
                        "fill": "#9e9e9e",
                      },
                    }
                  }
                  x1={0}
                  x2={400}
                  y1={0}
                  y2={0}
                />
                <g
                  className="rv-xy-plot__axis__ticks"
                  transform="translate(0, 0)"
                >
                  <g
                    className="rv-xy-plot__axis__tick"
                    style={
                      Object {
                        "fill": "#9e9e9e",
                        "line": Object {
                          "stroke": "none",
                        },
                        "text": Object {
                          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                          "fontSize": "0.875rem",
                        },
                        "ticks": Object {
                          "fill": "#9e9e9e",
                        },
                      }
                    }
                    transform="translate(75.94748676939308, 0)"
                  >
                    <line
                      className="rv-xy-plot__axis__tick__line"
                      style={
                        Object {
                          "fill": "#9e9e9e",
                          "line": Object {
                            "stroke": "none",
                          },
                          "stroke": "none",
                          "text": Object {
                            "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                            "fontSize": "0.875rem",
                          },
                          "ticks": Object {
                            "fill": "#9e9e9e",
                          },
                        }
                      }
                      x1={0}
                      x2={0}
                      y1={-0}
                      y2={6}
                    />
                    <text
                      className="rv-xy-plot__axis__tick__text"
                      dy="0.72em"
                      style={
                        Object {
                          "fill": "#9e9e9e",
                          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                          "fontSize": "0.875rem",
                          "line": Object {
                            "stroke": "none",
                          },
                          "text": Object {
                            "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                            "fontSize": "0.875rem",
                          },
                          "ticks": Object {
                            "fill": "#9e9e9e",
                          },
                        }
                      }
                      textAnchor="middle"
                      transform="translate(0, 14)"
                    >
                      1940
                    </text>
                  </g>
                  <g
                    className="rv-xy-plot__axis__tick"
                    style={
                      Object {
                        "fill": "#9e9e9e",
                        "line": Object {
                          "stroke": "none",
                        },
                        "text": Object {
                          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                          "fontSize": "0.875rem",
                        },
                        "ticks": Object {
                          "fill": "#9e9e9e",
                        },
                      }
                    }
                    transform="translate(158.6453886812003, 0)"
                  >
                    <line
                      className="rv-xy-plot__axis__tick__line"
                      style={
                        Object {
                          "fill": "#9e9e9e",
                          "line": Object {
                            "stroke": "none",
                          },
                          "stroke": "none",
                          "text": Object {
                            "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                            "fontSize": "0.875rem",
                          },
                          "ticks": Object {
                            "fill": "#9e9e9e",
                          },
                        }
                      }
                      x1={0}
                      x2={0}
                      y1={-0}
                      y2={6}
                    />
                    <text
                      className="rv-xy-plot__axis__tick__text"
                      dy="0.72em"
                      style={
                        Object {
                          "fill": "#9e9e9e",
                          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                          "fontSize": "0.875rem",
                          "line": Object {
                            "stroke": "none",
                          },
                          "text": Object {
                            "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                            "fontSize": "0.875rem",
                          },
                          "ticks": Object {
                            "fill": "#9e9e9e",
                          },
                        }
                      }
                      textAnchor="middle"
                      transform="translate(0, 14)"
                    >
                      1960
                    </text>
                  </g>
                  <g
                    className="rv-xy-plot__axis__tick"
                    style={
                      Object {
                        "fill": "#9e9e9e",
                        "line": Object {
                          "stroke": "none",
                        },
                        "text": Object {
                          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                          "fontSize": "0.875rem",
                        },
                        "ticks": Object {
                          "fill": "#9e9e9e",
                        },
                      }
                    }
                    transform="translate(241.3432905930075, 0)"
                  >
                    <line
                      className="rv-xy-plot__axis__tick__line"
                      style={
                        Object {
                          "fill": "#9e9e9e",
                          "line": Object {
                            "stroke": "none",
                          },
                          "stroke": "none",
                          "text": Object {
                            "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                            "fontSize": "0.875rem",
                          },
                          "ticks": Object {
                            "fill": "#9e9e9e",
                          },
                        }
                      }
                      x1={0}
                      x2={0}
                      y1={-0}
                      y2={6}
                    />
                    <text
                      className="rv-xy-plot__axis__tick__text"
                      dy="0.72em"
                      style={
                        Object {
                          "fill": "#9e9e9e",
                          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                          "fontSize": "0.875rem",
                          "line": Object {
                            "stroke": "none",
                          },
                          "text": Object {
                            "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                            "fontSize": "0.875rem",
                          },
                          "ticks": Object {
                            "fill": "#9e9e9e",
                          },
                        }
                      }
                      textAnchor="middle"
                      transform="translate(0, 14)"
                    >
                      1980
                    </text>
                  </g>
                  <g
                    className="rv-xy-plot__axis__tick"
                    style={
                      Object {
                        "fill": "#9e9e9e",
                        "line": Object {
                          "stroke": "none",
                        },
                        "text": Object {
                          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                          "fontSize": "0.875rem",
                        },
                        "ticks": Object {
                          "fill": "#9e9e9e",
                        },
                      }
                    }
                    transform="translate(324.0411925048147, 0)"
                  >
                    <line
                      className="rv-xy-plot__axis__tick__line"
                      style={
                        Object {
                          "fill": "#9e9e9e",
                          "line": Object {
                            "stroke": "none",
                          },
                          "stroke": "none",
                          "text": Object {
                            "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                            "fontSize": "0.875rem",
                          },
                          "ticks": Object {
                            "fill": "#9e9e9e",
                          },
                        }
                      }
                      x1={0}
                      x2={0}
                      y1={-0}
                      y2={6}
                    />
                    <text
                      className="rv-xy-plot__axis__tick__text"
                      dy="0.72em"
                      style={
                        Object {
                          "fill": "#9e9e9e",
                          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                          "fontSize": "0.875rem",
                          "line": Object {
                            "stroke": "none",
                          },
                          "text": Object {
                            "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
                            "fontSize": "0.875rem",
                          },
                          "ticks": Object {
                            "fill": "#9e9e9e",
                          },
                        }
                      }
                      textAnchor="middle"
                      transform="translate(0, 14)"
                    >
                      2000
                    </text>
                  </g>
                </g>
              </g>
            </svg>
          </div>
          <div
            className="MuiGrid-container-19 MuiGrid-spacing-xs-8-40"
          >
            <div
              className="MuiGrid-item-20"
            >
              <div
                aria-valuenow={85}
                className="MuiCircularProgress-root-169 MuiCircularProgress-colorPrimary-172 MuiCircularProgress-static-170"
                role="progressbar"
                style={
                  Object {
                    "display": "inline-box",
                    "height": 20,
                    "marginRight": "8px",
                    "transform": "rotate(-90deg)",
                    "verticalAlign": "middle",
                    "width": 20,
                  }
                }
              >
                <svg
                  className="MuiCircularProgress-svg-174"
                  viewBox="0 0 50 50"
                >
                  <circle
                    className="MuiCircularProgress-circle-175 MuiCircularProgress-circleStatic-176"
                    cx={25}
                    cy={25}
                    fill="none"
                    r={20}
                    strokeWidth={3.6}
                    style={
                      Object {
                        "strokeDasharray": "125.664",
                        "strokeDashoffset": "19.333px",
                      }
                    }
                  />
                </svg>
              </div>
              <p
                className="MuiTypography-root-144 MuiTypography-body1-153"
                style={
                  Object {
                    "display": "inline",
                  }
                }
              >
                84.6
                % moisture
              </p>
            </div>
            <div
              className="MuiGrid-item-20"
            >
              <div
                aria-valuenow={54}
                className="MuiCircularProgress-root-169 MuiCircularProgress-colorPrimary-172 MuiCircularProgress-static-170"
                role="progressbar"
                style={
                  Object {
                    "display": "inline-box",
                    "height": 20,
                    "marginRight": "8px",
                    "transform": "rotate(-90deg)",
                    "verticalAlign": "middle",
                    "width": 20,
                  }
                }
              >
                <svg
                  className="MuiCircularProgress-svg-174"
                  viewBox="0 0 50 50"
                >
                  <circle
                    className="MuiCircularProgress-circle-175 MuiCircularProgress-circleStatic-176"
                    cx={25}
                    cy={25}
                    fill="none"
                    r={20}
                    strokeWidth={3.6}
                    style={
                      Object {
                        "strokeDasharray": "125.664",
                        "strokeDashoffset": "57.428px",
                      }
                    }
                  />
                </svg>
              </div>
              <p
                className="MuiTypography-root-144 MuiTypography-body1-153"
                style={
                  Object {
                    "display": "inline",
                  }
                }
              >
                22.3°C ambient
              </p>
            </div>
            <div
              className="MuiGrid-item-20"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-178"
                color={undefined}
                focusable="false"
                style={
                  Object {
                    "verticalAlign": "middle",
                  }
                }
                viewBox="0 0 24 24"
              >
                <g>
                  <path
                    d="M16.01 7L16 3h-2v4h-4V3H8v4h-.01C7 6.99 6 7.99 6 8.99v5.49L9.5 18v3h5v-3l3.5-3.51v-5.5c0-1-1-2-1.99-1.99z"
                  />
                </g>
              </svg>
              <p
                className="MuiTypography-root-144 MuiTypography-body1-153"
                style={
                  Object {
                    "display": "inline",
                  }
                }
              >
                Pump running
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;
